### PR TITLE
Make 'ResourceOptions.merge' static so downstream clients don't have to manually handle None cases.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
I recommend reviewing with whitespace off.

### CODE REVIEW

File: sdk/python/lib/pulumi/resource.py [hunk 1]
```diff
@@ -342,57 +344,59 @@ def __init__(self,
...
 
-    def merge(self, other: 'ResourceOptions') -> 'ResourceOptions':
+    # pylint: disable=method-hidden
+    @staticmethod
+    def merge(opts1: 'ResourceOptions', opts2: 'ResourceOptions') -> 'ResourceOptions':
         """
-        merge produces a new ResourceOptions object with the respective attributes of this
-        instance in it with the attributes of `other` merged over them.
+        merge produces a new ResourceOptions object with the respective attributes of the
+        `opts` instance in it with the attributes of `opts2` merged over them.
```
Comment (CyrusNajmabadi) [inline sdk/python/lib/pulumi/resource.py]:
```suggestion
        `opts1` instance in it with the attributes of `opts2` merged over them.
```

File: sdk/python/lib/pulumi/resource.py [hunk 2]
```diff
@@ -342,57 +344,59 @@ def __init__(self,
...
-         4. Attributes with value 'None' will not be copied over.
+            1. if the attributes is a collection, the final value will be a collection containing
+               the values from each options object. Both original collections in each options
+               object will be unchanged.
+            2. Simple scaler values from `opts2` (i.e. strings, numbers, bools) will replace the
+               values from `opts1`.
+            3. For the purposes of merging `depends_on`, `provider` and `providers` are always
+               treated as collections, even if only a single value was provided.
+            4. Attributes with value 'None' will not be copied over.
         """
```
Comment (CyrusNajmabadi) [inline sdk/python/lib/pulumi/resource.py]:
```suggestion
        This method can be called either as static-method like `ResourceOptions.merge(opts1, opts2)` or as an instance-method
        like `opts1.merge(opts2)`.  The former is useful for cases where `opts1` may be `None` so the caller does not need
        to check for this case. 
        """
```

File: sdk/python/lib/pulumi/resource.py [hunk 3]
```diff
@@ -342,57 +344,62 @@ def __init__(self,
...
+    @staticmethod
+    def merge(opts1: 'ResourceOptions', opts2: 'ResourceOptions') -> 'ResourceOptions':
         """
-        merge produces a new ResourceOptions object with the respective attributes of this
-        instance in it with the attributes of `other` merged over them.
+        merge produces a new ResourceOptions object with the respective attributes of the
+        `opts1` instance in it with the attributes of `opts2` merged over them.
 
-        Both this options instance and the `other` options instance will be unchanged.
+        Both the `opts1` instance and the `opts2` instance will be unchanged.
```
Comment (CyrusNajmabadi) [inline sdk/python/lib/pulumi/resource.py]:
```suggestion
        Both the `opts1` instance and the `opts2` instance will be unchanged.  Both of `opts1` and `opts2` can be 
        `None`, in which case its attributes are ignored.
```

File: sdk/python/lib/pulumi/resource.py [hunk 4]
```diff
@@ -342,57 +344,63 @@ def __init__(self,
...
-         4. Attributes with value 'None' will not be copied over.
+            1. if the attributes is a collection, the final value will be a collection containing
+               the values from each options object. Both original collections in each options
+               object will be unchanged.
+            2. Simple scaler values from `opts2` (i.e. strings, numbers, bools) will replace the
+               values from `opts1`.
+            3. For the purposes of merging `depends_on`, `provider` and `providers` are always
+               treated as collections, even if only a single value was provided.
+            4. Attributes with value 'None' will not be copied over.
+        This method can be called either as static-method like `ResourceOptions.merge(opts1, opts2)` or as an instance-method
```
Comment (CyrusNajmabadi) [inline sdk/python/lib/pulumi/resource.py]:
```suggestion

        This method can be called either as static-method like `ResourceOptions.merge(opts1, opts2)` or as an instance-method
```

File: sdk/python/.pylintrc [hunk 5]
```diff
@@ -159,7 +159,8 @@ disable=print-statement,
         broad-except,
         no-self-use,
         unused-import,
-        unsubscriptable-object
+        unsubscriptable-object,
+        line-too-long
```
Comment (lukehoban) [inline sdk/python/.pylintrc]: This feels like one we may not want to turn off - I think we do for the most part want to maintain discipline around this?

File: sdk/python/lib/pulumi/resource.py [hunk 6]
```diff
@@ -342,57 +344,68 @@ def __init__(self,
                     raise Exception(
                         "'depends_on' was passed a value that was not a Resource.")
 
+    def _merge_instance(self, opts: 'ResourceOptions') -> 'ResourceOptions':
```
Comment (lukehoban) [inline sdk/python/lib/pulumi/resource.py]: How do doc comments work in this situation?  Do we need to include doc comments on this `_merge_instance` that will get picked up and displayed for the instance `merge`?

File: sdk/python/.pylintrc (hunk 5 above)
Comment (CyrusNajmabadi) [reply sdk/python/.pylintrc]: i find this rule just enormous frustrating.  At 100, it feels like even basic stuff just constantly hits it.  It's particularly annoying in libraries where we're trying to keep consistency.  in this case, i have a bunch of statements in a row that would be good to keep totally in sync, having to wrap one at the expense of the others just hurts readability for me.  

I feel like we can trust people to make reasonable judgements on line length TBH.

File: sdk/python/lib/pulumi/resource.py (hunk 6 above)
Comment (CyrusNajmabadi) [reply sdk/python/lib/pulumi/resource.py]: It will not.  But i've updated this PR to rectify that.  Good catch!

**Pull request merged**